crates/flox-rust-sdk/src/models/environment/generations.rs

```diff
@@ -0,0 +1,360 @@
+use flox_types::version::Version;
+use serde::{Deserialize, Serialize};
+use serde_with::{DeserializeFromStr, SerializeDisplay};
+use thiserror::Error;
+
+use super::path_environment::{Original, PathEnvironment};
+use super::{copy_dir_recursive, PathPointer, ENV_DIR_NAME};
+use crate::providers::git::{GitCommandProvider, GitProvider};
+
+const GENERATIONS_METADATA_FILE: &str = "metadata.json";
+
+pub struct ReadOnly {}
+
+pub struct ReadWrite {
+    /// A base directory that contains the tempdir
+    /// of the checked out generations branch
+    /// as well as tempdirs created by [PathEnvironment] transactions.
+    tempdir_base: PathBuf,
+}
+
+/// A representation of the generations of an environment
+///
+/// Example File layout:
+///
+/// ./
+/// ├── 1
+/// │  └── env
+/// │     ├── manifest.toml
+/// │     └── manifest.lock
+/// ├── 2
+/// │  └── env
+/// │    └── manifest.toml (lockfile is optional)
+/// ├── ... N
+/// │  └── env
+/// │     └── manifest.toml
+/// └── metadata.json
+pub struct Generations<State> {
+    /// A floxmeta repository/branch that contains the generations of an environment
+    ///
+    /// - When [ReadOnly], this is assumend to be bare, but it is not enforced.
+    /// - When [ReadWrite], this is required to not be bare.
+    ///   This is enforced when created using [Self::writable].
+    repo: GitCommandProvider,
+
+    /// A path pointer for the environment that will
+    /// be associated with a realized generation
+    pointer: PathPointer,
+
+    branch: String,
+
+    state: State,
+}
+
+impl<S> Generations<S> {
+    /// Read the generations metadata for an environment
+    pub fn metadata(&self) -> Result<Metadata, GenerationsError> {
+        read_metadata(&self.repo, &self.branch)
+    }
+}
+
+impl Generations<ReadOnly> {
+    /// Create a new generations instance
+    pub fn new(
+        environment_repr: GitCommandProvider,
+        ref_name: String,
+        pointer: PathPointer,
+    ) -> Self {
+        Self {
+            repo: environment_repr,
+            branch: ref_name,
+            pointer,
+            state: ReadOnly {},
+        }
+    }
+
+    pub fn writable(
+        self,
+        tempdir_base: PathBuf,
+    ) -> Result<Generations<ReadWrite>, GenerationsError> {
+        let repo = checkout_to_tempdir(
+            &self.repo,
+            &self.branch,
+            tempfile::tempdir_in(&tempdir_base).unwrap().into_path(),
+        )?;
+
+        Ok(Generations {
+            repo,
+            branch: self.branch,
+            pointer: self.pointer,
+            state: ReadWrite { tempdir_base },
+        })
+    }
+}
+
+impl Generations<ReadWrite> {
+    /// Realize a generation as a [PathEnvironment]
+    ///
+    /// The generation can then be safely modified
+    /// and registered as a new generation using [Self::add_generation].
+    pub fn realize_generation(
+        &self,
+        generation: usize,
+    ) -> Result<PathEnvironment<Original>, GenerationsError> {
```

Comment:
note/suggestion blocking: since this returns a PathEnvironment and based on the other PR, I'm expecting you're planning to use PathEnvironment methods like PathEnvironment::install(). That will create the gcroot in the temporary git directory rather than floxmeta.

I'm not sure what the solution to that is. We could:
1. Leave things close to how you have, don't check in the `run` directory, and add a gcroot manually after the fact. That feels pretty hacky 
2. Create a shared inner environment that is just the `env` directory. That feels the cleanest.
3. Implement methods like install completely separately for ManagedEnvironment for now. That feels clean enough and fast enough 

I think you should feel free to decide to go with whatever you think is best if you're working on this

Reply:
I'll address this in the other PR.

Right now I think more of a solution like (1) and possibly (2) later on.

I like 2, it's really just separating the modifying env operations into a sngle place and lets the concrete environments just deal with the differences..
I'd prefer that, but for now, modifying `build` to take a out-link path and "building" a second time to that path seems the quickest way forward given our time constraints.